src/libraries/System.Private.CoreLib/src/System/ComAwareWeakReference.cs

```diff
@@ -0,0 +1,179 @@
+            private readonly long _wrapperId;
+
+            internal object? ResolveTarget()
+            {
+                return ComWeakRefToObject(_pComWeakRef, _wrapperId);
+            }
+
+            // see: syncblk.h
+            private const int IS_HASHCODE_BIT_NUMBER = 26;
+            private const int BIT_SBLK_IS_HASHCODE = 1 << IS_HASHCODE_BIT_NUMBER;
+            private const int BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX = 0x08000000;
+
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            internal static unsafe ComInfo? FromObject(object? target)
+            {
+                if (target == null)
+                    return null;
+
+                fixed (byte* pRawData = &target.GetRawData())
+                {
+                    // The header is 4 bytes before MT field on all architectures
+                    int header = *(int*)(pRawData - sizeof(IntPtr) - sizeof(int));
+                    // common case: target does not have a syncblock, so there is no interop info
+                    if ((header & (BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX | BIT_SBLK_IS_HASHCODE)) != BIT_SBLK_IS_HASH_OR_SYNCBLKINDEX)
+                        return null;
+                }
+
+                return FromObjectSlow(target);
+            }
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            private static unsafe ComInfo? FromObjectSlow(object target)
+            {
+                IntPtr pComWeakRef = ObjectToComWeakRef(target, out long wrapperId);
+                if (pComWeakRef == 0)
+                    return null;
+
+                return new ComInfo(pComWeakRef, wrapperId);
+            }
+
+            private ComInfo(IntPtr pComWeakRef, long wrapperId)
+            {
+                Debug.Assert(pComWeakRef != IntPtr.Zero);
+                _pComWeakRef = pComWeakRef;
+                _wrapperId = wrapperId;
+            }
+
+            ~ComInfo()
+            {
+                Marshal.Release(_pComWeakRef);
+            }
+        }
+
+        private ComAwareWeakReference(nint weakHandle)
+        {
+            Debug.Assert(weakHandle != 0);
+            _weakHandle = weakHandle;
+        }
+
+        ~ComAwareWeakReference()
+        {
+            GCHandle.InternalFree(_weakHandle);
+        }
+
+        private void SetTarget(object? target, ComInfo? comInfo)
+        {
+            // NOTE: ComAwareWeakReference is an internal implementation detail and
+            //       instances are never exposed publicly, thus we can use "this" for locking
+            lock (this)
+            {
+                GCHandle.InternalSet(_weakHandle, target);
+                _comInfo = comInfo;
+            }
+        }
+
+        internal object? Target => GCHandle.InternalGet(_weakHandle) ?? RehydrateTarget();
+
+        private object? RehydrateTarget()
+        {
+            object? target = null;
+            lock (this)
+            {
+                if (_comInfo != null)
```

Comment:
As a fast path, we can check for `if (_comInfo != null)` before taking the lock as well.

Reply:
I think we can only return null here if we see both target and _comInfo as not referring to an instance and we can only tell that while holding a lock. Otherwise if a COM target with a dead RCW is replaced by an ordinary non-COM target, a concurrent reader may return null, while the reference was continuously referring to something alive.

I think getting here will be a very rare scenario though, and since this is a private instance and noone does GetHashcode, the lock should be relatively cheap, so perhaps not a big perf concern.

Reply:
> Otherwise if a COM target with a dead RCW is replaced by an ordinary non-COM target, a concurrent reader may return null, while the reference was continuously referring to something alive.

Good point.